src/librustc_incremental/persist/load.rs

```diff
@@ -168,35 +170,40 @@ pub fn load_dep_graph(sess: &Session) ->
-                LoadResult::DataOutOfDate => LoadResult::DataOutOfDate,
-                LoadResult::Error { message } => LoadResult::Error { message },
-                LoadResult::Ok { data: (bytes, start_pos) } => {
-
-                    let mut decoder = Decoder::new(&bytes, start_pos);
-                    let prev_commandline_args_hash = u64::decode(&mut decoder)
-                        .expect("Error reading commandline arg hash from cached dep-graph");
-
-                    if prev_commandline_args_hash != expected_hash {
-                        if report_incremental_info {
-                            println!("[incremental] completely ignoring cache because of \
-                                    differing commandline arguments");
-                        }
-                        // We can't reuse the cache, purge it.
-                        debug!("load_dep_graph_new: differing commandline arg hashes");
-
-                        // No need to do any further work
-                        return LoadResult::DataOutOfDate;
-                    }
+    let load = move || {
+        match load_data(report_incremental_info, &path) {
+            LoadResult::DataOutOfDate => LoadResult::DataOutOfDate,
+            LoadResult::Error { message } => LoadResult::Error { message },
+            LoadResult::Ok { data: (bytes, start_pos) } => {
+
+                let mut decoder = Decoder::new(&bytes, start_pos);
+                let prev_commandline_args_hash = u64::decode(&mut decoder)
+                    .expect("Error reading commandline arg hash from cached dep-graph");
 
-                    let dep_graph = SerializedDepGraph::decode(&mut decoder)
-                        .expect("Error reading cached dep-graph");
+                if prev_commandline_args_hash != expected_hash {
+                    if report_incremental_info {
+                        println!("[incremental] completely ignoring cache because of \
+                                differing commandline arguments");
+                    }
+                    // We can't reuse the cache, purge it.
+                    debug!("load_dep_graph_new: differing commandline arg hashes");
 
-                    LoadResult::Ok { data: PreviousDepGraph::new(dep_graph) }
+                    // No need to do any further work
+                    return LoadResult::DataOutOfDate;
                 }
+
+                let dep_graph = SerializedDepGraph::decode(&mut decoder)
+                    .expect("Error reading cached dep-graph");
+
+                LoadResult::Ok { data: PreviousDepGraph::new(dep_graph) }
             }
+        }
+    };
+
+    MaybeAsync::Async(std::thread::spawn(move || {
+        time_ext(time_passes, None, "background load prev dep-graph", move || {
+            panic::catch_unwind(load).unwrap_or_else(
+                |_| LoadResult::Error { message: "could not decode incremental cache".into() })
```

Comment:
Please emit the actual panic via a debug! or even info! call, so we can figure out what's going on without needing to rebuild the compiler.

Reply:
I couldn't figure out a way to do this. I tried doing formatting the error into the message with `format!("could not decode incremental cache: {:?}", e)` but that just printed `"could not decode incremental cache: Any"`.

Reply:
Maybe someone for @rust-lang/libs knows how to do this best?

Reply:
There's currently no easy way to get out the message of a `Box<Any>` unfortunately, but we also tend to recommend against using `catch_unwind` for error handling like this and would instead recommend returning an `Err` all the way up (which can then be printed)

Reply:
The panic handler for the thread is still running so the compiler output looks like this:

```
   Compiling webrender v0.57.0 (file:///Users/wesley/code/rust/temp3/webrender/webrender)
thread '<unnamed>' panicked at 'internal error: entered unreachable code', librustc/dep_graph/dep_node.rs:124:34
note: Run with `RUST_BACKTRACE=1` for a backtrace.
warning: could not decode incremental cache
```